pkg/controller/tainteviction/timed_workers.go

```diff
@@ -54,35 +55,54 @@ type TimedWorker struct {
 	CreatedAt time.Time
 	FireAt    time.Time
 	Timer     clock.Timer
+	cancelled atomic.Bool
 }
 
 // createWorker creates a TimedWorker that will execute `f` not earlier than `fireAt`.
 // Returns nil if the work was started immediately and doesn't need a timer.
-func createWorker(ctx context.Context, args *WorkArgs, createdAt time.Time, fireAt time.Time, f func(ctx context.Context, fireAt time.Time, args *WorkArgs) error, clock clock.WithDelayedExecution) *TimedWorker {
+func createWorker(ctx context.Context, wg *sync.WaitGroup, args *WorkArgs, createdAt time.Time, fireAt time.Time, f func(ctx context.Context, fireAt time.Time, args *WorkArgs) error, clock clock.WithDelayedExecution) *TimedWorker {
 	delay := fireAt.Sub(createdAt)
 	logger := klog.FromContext(ctx)
-	fWithErrorLogging := func() {
-		err := f(ctx, fireAt, args)
-		if err != nil {
-			logger.Error(err, "TaintEvictionController: timed worker failed")
-		}
-	}
-	if delay <= 0 {
-		go fWithErrorLogging()
-		return nil
-	}
-	timer := clock.AfterFunc(delay, fWithErrorLogging)
-	return &TimedWorker{
+
+	worker := TimedWorker{
 		WorkItem:  args,
 		CreatedAt: createdAt,
 		FireAt:    fireAt,
-		Timer:     timer,
 	}
+
+	// This dance with the cancelled flag is here so that we can be sure that once TimedWorker.Cancel returns,
+	// we either never get to any processing, or the thread is already registered with the WaitGroup.
+	// Otherwise, the following sequence can happen:
+	//   1. TimedWorker.Timer fires and gets to go wrapper().
+	//   2. TimedWorker.Cancel is called to stop the timer, but a goroutine is already running.
+	//      It's started, but wg.Go hasn't been called yet to register the goroutine.
+	//   3. We call wg.Wait, which unblocks, because the inner wg.Go hasn't been called yet.
+	//      This causes wg.Wait to unblock after TimedWorker.Cancel is called and still start a goroutine.
+	// So in our case we can still get to starting an unregistered goroutine, but it will exit immediately.
+	wrapper := func() {
+		wg.Go(func() {
+			if worker.cancelled.Load() {
+				return
+			}
```

Comment:
I am really not in love with this, but feel free to propose a different solution.